### PR TITLE
New Functionality to arrays - arrayRemove

### PR DESCRIPTION
**Pull request by rajatmohan22**
### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

A new function on arrays called ArrayRemove. Its working is as follows: arayRemove(x -> cond, arr), This removes those elements in the array for which the condition is true.
Documentation entry for user-facing changes

- [x] Documentation is written (mandatory for new features)

Motivation: This function aims to address certain confusions caused by arrayFilter. The arrayFilter function operates mysteriously using the syntax arrayRemove(x -> cond, arr), where the elements that satisfy the condition are retained in the array. This can lead to confusion for users. To simplify and clarify the process, we introduce the arrayRemove function, which works directly as its name suggests. It removes the elements that pass the given condition, offering a more intuitive and straightforward approach.

parameters: :
-> arrayRemove(arr, elem).
arr refers to the array and elem refers to the element to be removed.

-> Works with lambda functions as well:
arrayRemove(arr, x-> cond)
arr refers to the array. the second parameter refers to the condition by which the element must be removed.

Example:

SELECT
arrayRemove(x -> x % 2 = 0,[2,3,4,5,6]) AS example

Link to an older closed PR: https://github.com/ClickHouse/ClickHouse/pull/52371

### CODE REVIEW

**Comment by rajatmohan22** [comment]
Can someone please help me out?
Why is that this error occurs?

Received exception from server (version 23.8.1):
Code: 49. DB::Exception: Received from localhost:9000. DB::Exception: offsets_column has data inconsistent with nested_column. Data size: 1, last offset: 3: while executing 'FUNCTION arrayRemove(__lambda :: 2, [2, 10, 11, 12] :: 0) -> arrayRemove(lambda(tuple(x), less(x, 3)), [2, 10, 11, 12]) Array(UInt8) : 1'. (LOGICAL_ERROR)

**Comment by nickitat** [comment]
the purpose of the second function like `arrayFilter` is unclear tbh. isn't `arrayRemove(arr, cond)` ≡ `arrayFilter(arr, !cond)`?

**Comment by alexey-milovidov** [comment]
@nickitat Yes, this is only for consistency - because it's natural to expect that the function also works in this way.

**Comment by nickitat** [comment]
ok, so @rajatmohan22 let's reuse implementation of `arrayFilter`?

**Comment by rajatmohan22** [comment]
@nickitat I've done exactly that.
Array Filter:

`for (size_t i = 0; i < in_offsets.size(); ++i)
    {
        for (; in_pos < in_offsets[i]; ++in_pos)
        {
            if (filter[in_pos])
                ++out_pos;
        }
        out_offsets[i] = out_pos;
    }`

Array Remove:

`for (size_t i = 0; i < in_offsets.size(); ++i)
    {
        for (; in_pos < in_offsets[i]; ++in_pos)
        {
            if (!filter[in_pos])
                ++out_pos;
        }
        out_offsets[i] = out_pos;
    }`

**Comment by rajatmohan22** [comment]
/home/rajat/Pictures/Screenshots/Screenshot from 2023-07-31 06-53-20.png

File: tests/queries/0_stateless/02415_all_new_functions_must_be_documented.reference [hunk 1]
```diff
@@ -127,6 +127,7 @@ arrayPushBack
 arrayPushFront
 arrayReduce
 arrayReduceInRanges
+arrayRemove
```
Comment (alexey-milovidov) [inline tests/queries/0_stateless/02415_all_new_functions_must_be_documented.reference:130]: All new functions must be documented inline in the code, see `asin.cpp`

File: docs/en/sql-reference/functions/array-functions.md [hunk 2]
```diff
@@ -1794,6 +1794,52 @@ Return value type is always [Float64](../../sql-reference/data-types/float.md).
...
+**Syntax**
+
+``` sql
+arrayRemove(array, predicate)
+```
+
+**Arguments**
+
+- `array` – Array.
+- `predicate` – Condition to remove elements from the array
```
Comment (alexey-milovidov) [inline docs/en/sql-reference/functions/array-functions.md:1810]: It should work without a predicate as well:

`arrayRemove(arr, value)`

**Comment by alexey-milovidov** [comment]
Need to add functional tests.

File: docs/en/sql-reference/functions/array-functions.md (hunk 2 above)
Comment (rajatmohan22) [reply docs/en/sql-reference/functions/array-functions.md:1810]: I've kept the working very similar to arrayFilter. Maybe we can define another function for that sort of a functioningn such as `arrayRemoveElement` ?

**Comment by rajatmohan22** [comment]
> Need to add functional tests.

added

**Comment by rajatmohan22** [comment]
Can someone review this ince? 273 tests are successful.

File: docs/en/sql-reference/functions/array-functions.md [hunk 3]
```diff
@@ -1794,6 +1794,55 @@ Return value type is always [Float64](../../sql-reference/data-types/float.md).
 │ 6   │ Float64                                                                                  │
 └─────┴──────────────────────────────────────────────────────────────────────────────────────────┘
 ```
+## arrayRemove
```
Comment (alexey-milovidov) [inline docs/en/sql-reference/functions/array-functions.md:1797]: It does not describe the usual way to call the function:

arrayRemove(element, array)

File: src/Functions/array/arrayRemove.cpp [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+                array.getDataPtr()->cloneEmpty(),
+                ColumnArray::ColumnOffsets::create(array.size(), 0));
+    }
+
+    const IColumn::Filter & filter = column_filter->getData();
+    IColumn::Filter negated_filter(filter.size()); // Create a new filter with the negated values, handling NULLs separately.
+
+    for (size_t i = 0; i < filter.size(); ++i)
+    {
+            negated_filter[i] = !filter[i];
```
Comment (alexey-milovidov) [inline src/Functions/array/arrayRemove.cpp:36]: Bad indent.

File: src/Functions/array/arrayRemove.cpp [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+        else
+            return ColumnArray::create(
+                array.getDataPtr()->cloneEmpty(),
+                ColumnArray::ColumnOffsets::create(array.size(), 0));
+    }
+
+    const IColumn::Filter & filter = column_filter->getData();
+    IColumn::Filter negated_filter(filter.size()); // Create a new filter with the negated values, handling NULLs separately.
+
+    for (size_t i = 0; i < filter.size(); ++i)
```
Comment (alexey-milovidov) [inline src/Functions/array/arrayRemove.cpp:34]: size is calculated in a loop

File: src/Functions/array/arrayRemove.cpp [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int ILLEGAL_COLUMN;
+}
+
+ColumnPtr ArrayRemoveImpl::execute(const ColumnArray & array, ColumnPtr mapped)
```
Comment (alexey-milovidov) [inline src/Functions/array/arrayRemove.cpp:12]: I don't see how the normal mode of operation:

```
arrayRemove(array, element)
```

could work.

File: tests/queries/0_stateless/02415_all_new_functions_must_be_documented.reference (hunk 1 above)
Comment (alexey-milovidov) [reply tests/queries/0_stateless/02415_all_new_functions_must_be_documented.reference:130]: All new functions must be documented inline in the code, see `asin.cpp`

File: tests/queries/0_stateless/02841_array_remove.sql [hunk 7]
```diff
@@ -0,0 +1,2 @@
+SELECT length(arrayRemove(x->(x%2==0), [1,2,3,4,5,6,7,8,9,10])) AS x;
```
Comment (alexey-milovidov) [inline tests/queries/0_stateless/02841_array_remove.sql:1]: Missing tests for the normal mode of operation:

```
arrayRemove(array, element)
```

and the compatibility alias for PostgreSQL

**Comment by alexey-milovidov** [comment]
I've reviewed it, and this PR does not (yet) implement the specification: https://github.com/ClickHouse/ClickHouse/issues/52099 A few changes are required.

File: docs/en/sql-reference/functions/array-functions.md (hunk 2 above)
Comment (alexey-milovidov) [reply docs/en/sql-reference/functions/array-functions.md]: We should do it exactly as in the specification: one function arrayRemove, two overloads, and a compatibility alias for Postgres.

File: src/Functions/array/arrayRemove.cpp (hunk 6 above)
Comment (rajatmohan22) [reply src/Functions/array/arrayRemove.cpp:12]: There are 3 things that can be done for this:

1. Overload the function arrayRemove.
2. add another argument in the same function as size_t element to remove ( or we could do index to remove from - easier complexity wise ).
3. add another function called arrayRemoveElememt: which does -> `arrayRemoveElement(arr,ele)`

File: src/Functions/array/arrayRemove.cpp (hunk 5 above)
Comment (rajatmohan22) [reply src/Functions/array/arrayRemove.cpp:34]: the` for loop` isn't calculating the size, it is creating the resultant array with the negative offsets.
i tried out a lot of things and spoke to @nickitat too, this was the only method which worked :(

File: src/Functions/array/arrayRemove.cpp (hunk 4 above)
Comment (rajatmohan22) [reply src/Functions/array/arrayRemove.cpp:36]: will fix that

**Comment by rajatmohan22** [comment]
Need help here! I've been trying from long but i do not know what is going wrong with the 2nd overload function. Could someone please help

File: docs/en/sql-reference/functions/array-functions.md.bak [hunk 8]
```diff
@@ -0,0 +1,1849 @@
+---
```
Comment (alexey-milovidov) [inline docs/en/sql-reference/functions/array-functions.md.bak:1]: This is committed by mistake.

**Comment by alexey-milovidov** [comment]
@nickitat is on vacation, he will check later.

File: tests/queries/0_stateless/02841_array_remove.sql (hunk 7 above)
Comment (nickitat) [reply tests/queries/0_stateless/02841_array_remove.sql:1]: * pls add test Alexey asked for
* to learn how to add an alias check this [pr](https://github.com/ClickHouse/ClickHouse/commit/290c145acfc5accb2344b63356904af5f0a784ba)

File: src/Functions/array/arrayRemove.cpp [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+        out_offsets[i] = out_pos;
+    }
+
+    return ColumnArray::create(filtered, std::move(column_offsets));
+
+}
+
+REGISTER_FUNCTION(ArrayRemove)
+{
+    factory.registerFunction<FunctionArrayRemove>();
```
Comment (nickitat) [inline src/Functions/array/arrayRemove.cpp:100]: pls add inline docs like this:
https://github.com/ClickHouse/ClickHouse/blob/45c28e1221fa9f3903318bc85e19f5cbdd93c448/src/Functions/asin.cpp#L17-L45

**Review by nickitat — CHANGES_REQUESTED** [review]

**Comment by alexey-milovidov** [comment]
@rajatmohan22 please merge with master.

**Comment by alexey-milovidov** [comment]
@rajatmohan22, please merge with the upstream repository's master.

File: tests/queries/0_stateless/02841_array_remove.sql (hunk 7 above)
Comment (alexey-milovidov) [reply tests/queries/0_stateless/02841_array_remove.sql:1]: @rajatmohan22, need to add the required overload.

File: tests/queries/0_stateless/02415_all_new_functions_must_be_documented.reference (hunk 1 above)
Comment (alexey-milovidov) [reply tests/queries/0_stateless/02415_all_new_functions_must_be_documented.reference:130]: All new functions must be documented inline in the code, see `asin.cpp`

**Comment by alexey-milovidov** [comment]
This PR looks abandoned.

**Comment by Algunenano** [comment]
This PR is abandoned

**Pull request closed without merging**